src/fs.cpp

```diff
@@ -52,15 +55,35 @@ bool FileLock::TryLock()
     if (fd == -1) {
         return false;
     }
-    struct flock lock;
-    lock.l_type = F_WRLCK;
-    lock.l_whence = SEEK_SET;
-    lock.l_start = 0;
-    lock.l_len = 0;
-    if (fcntl(fd, F_SETLK, &lock) == -1) {
-        reason = GetErrorReason();
-        return false;
+
+    // Exclusive file locking is broken on WSL using fcntl (issue #18622)
+    // This workaround can be removed once the bug on WSL is fixed
+    // Cache the result to avoid multiple system calls
+    static bool is_wsl_cache_exists = false;
```

Comment:
It's a crazy edge case, and unlikely on realistic platforms, but this is technically UB if two threads simultaneously reach the `if (!is_wsl_cache_exists)` check. The most C++ish solution is this I think: https://github.com/sipa/bitcoin/commit/9470c59e7290774972f94250d04caae161bc7678.